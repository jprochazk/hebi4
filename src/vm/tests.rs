use std::{fs::read_to_string, path::Path};

#[glob_test::glob("../../tests/inputs/run/*.hi")]
fn run(path: &Path) {
    let input = read_to_string(path).unwrap();
    let tokens = crate::token::tokenize(&input);
    let ast = match crate::parser::parse(&tokens) {
        Ok(ast) => ast,
        Err(err) => {
            panic!("{}", err.render(&input));
        }
    };
    let chunk = match crate::codegen::emit(&ast) {
        Ok(chunk) => chunk,
        Err(err) => panic!("{}", err.render(&input)),
    };
    let disasm = chunk.disasm(&input).to_string();

    let mut vm = crate::vm::Vm::new();
    vm.with(|mut r| {
        let (snapshot, failure) = match r.run(chunk) {
            Ok(value) => (format!("OK\n{}", r.fmt(value)), false),
            Err(err) => (
                format!("ERROR\n{}\n\nDISASM:\n\n{disasm}", err.render(&input)),
                true,
            ),
        };
        #[cfg(not(miri))]
        {
            insta::assert_snapshot!(snapshot);
            let _ = failure;
        }

        #[cfg(miri)]
        {
            if failure {
                panic!("{snapshot}");
            } else {
                eprintln!("{snapshot}");
            }
        }
    });
}
