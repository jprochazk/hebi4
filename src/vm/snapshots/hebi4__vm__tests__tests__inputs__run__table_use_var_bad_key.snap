---
source: src/vm/tests.rs
expression: snapshot
---
ERROR
value is not a valid table key
3 |  v[k]
  |   ^^^


DISASM:

.fn @main
  params 0
  stack 3
  .locals
    v = r1
    k = r2
  .code
  0  ltable r1, 1
  1  lsmi r2, 100
  2  skeyc r1, "a", r2
  3  lsmi r2, 0
  4  lidx r0, r1, r2
  5  ret
